contrib/firebaseio/src/main/java/io/FirebaseSource.java

```diff
@@ -0,0 +1,416 @@
+          }
+          ref.addListenerForSingleValueEvent(new ValueEventListener(){
+            @Override
+            public void onCancelled(FirebaseError err) {
+              error = err;
+            }
+            @Override
+            public void onDataChange(DataSnapshot timestamp) {
+              checkpoint.put(new Record<T>(
+                  new Instant(timestamp.getValue(Long.class)),
+                  event));
+            }
+          });
+        }
+      });
+    }
+
+    /**
+     * @see com.google.cloud.dataflow.sdk.io.UnboundedSource.UnboundedReader#advance()
+     **/
+    @Override
+    public boolean advance() throws IOException {
+      if (error == null){
+        return this.checkpoint.advance();
+      }
+      throw error.toException();
+    }
+
+    /**
+     * @see com.google.cloud.dataflow.sdk.io.UnboundedSource.UnboundedReader#getCheckpointMark()
+     **/
+    @Override
+    public CheckpointMark getCheckpointMark() {
+      return checkpoint;
+    }
+
+    /**
+     * @see com.google.cloud.dataflow.sdk.io.UnboundedSource.UnboundedReader#getCurrentRecordId()
+     **/
+    @Override
+    public byte[] getCurrentRecordId() throws NoSuchElementException {
+      return digest.digest(
+          ByteBuffer.allocate(12).putInt(
+              this.checkpoint.read().data.hashCode()).putLong(
+                  this.checkpoint.read().timestamp.getMillis()).array());
+    }
+
+    /**
+     * @see com.google.cloud.dataflow.sdk.io.UnboundedSource.UnboundedReader#getCurrentSource()
+     **/
+    @Override
+    public UnboundedSource<FirebaseEvent<T>, ?> getCurrentSource() {
+      return this.source;
+    }
+
+    /**
+     * @see com.google.cloud.dataflow.sdk.io.UnboundedSource.UnboundedReader#getWatermark()
+     **/
+    @Override
+    public Instant getWatermark() {
+      return this.checkpoint.minTimestamp();
+    }
+
+    /**
+     * @see com.google.cloud.dataflow.sdk.io.UnboundedSource.UnboundedReader#start()
+     **/
+    @Override
+    public boolean start() throws IOException {
+      try {
+        this.source.auther.authenticate(this.source.query.getRef());
+      } catch (Exception e) {
+        throw new IOException(e);
+      }
+      if (this.source.listenForChildEvents){
+        this.childListener = this.source.query.addChildEventListener(this.childListener);
+      }
+      if (this.source.listenForValueEvents){
+        this.valueListener = this.source.query.addValueEventListener(this.valueListener);
+      }
+      return this.advance();
+    }
+
+    /**
+     * @see com.google.cloud.dataflow.sdk.io.Source.Reader#close()
+     **/
+    @Override
+    public void close() throws IOException {
+      if (this.source.listenForChildEvents) {
+        this.source.query.removeEventListener(this.childListener);
+      }
+      if (this.source.listenForValueEvents){
+        this.source.query.removeEventListener(this.valueListener);
+      }
+      final Semaphore lock = new Semaphore(1);
```

Comment:
Hmmm not sure. The Firebase Client Library uses non-standard asynchronous constructs (completion listener is from their library) so I'm not sure I would be able to wrap that option up in a future (I'm assuming you mean using `Future.await()` instead of a nasty semaphore business.